pkg/pillar/cmd/volumemgr/create.go

```diff
@@ -33,55 +34,82 @@ func createVdiskVolume(ctx *volumemgrContext, status types.VolumeStatus,
 
 	created := false
 
+	persistFsType := ctx.persistType
+
 	// this is the target location, where we expect the volume to be
 	filelocation := status.PathName()
-	if _, err := os.Stat(filelocation); err == nil {
-		errStr := fmt.Sprintf("Can not create %s for %s: exists",
-			filelocation, status.Key())
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-
-	// use the edge-containers library to extract the data we need
-	puller := registry.Puller{
-		Image: ref,
-	}
-
-	casClient, err := cas.NewCAS(casClientType)
-	if err != nil {
-		err = fmt.Errorf("Run: exception while initializing CAS client: %s", err.Error())
-		return created, "", err
-	}
-	defer casClient.CloseClient()
-	ctrdCtx, done := casClient.CtrNewUserServicesCtx()
-	defer done()
 
-	resolver, err := casClient.Resolver(ctrdCtx)
-	if err != nil {
-		errStr := fmt.Sprintf("error getting CAS resolver: %v", err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-
-	// create a writer for the file where we want
-	f, err := os.Create(filelocation)
-	if err != nil {
-		errStr := fmt.Sprintf("error creating target file at %s: %v", filelocation, err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-	defer f.Close()
-
-	if _, _, err := puller.Pull(&registry.FilesTarget{Root: f, AcceptHash: true}, 0, false, os.Stderr, resolver); err != nil {
-		errStr := fmt.Sprintf("error pulling %s from containerd: %v", ref, err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-
-	// Do we need to expand disk?
-	if err := maybeResizeDisk(filelocation, status.MaxVolSize); err != nil {
-		log.Error(err)
-		return created, "", err
+	if persistFsType != types.PersistZFS {
```

Comment:
Looks like separate functions for ZFS and non-ZFS paths would make this code look better.

Reply:
+1

Reply:
Well, I can concentrate on minimizing of checks like `persistFsType != types.PersistZFS` in favor of checking `info.Mode()&os.ModeDevice` of files. I think, It is not good idea to have lots of functions with zfs in name here, the good way is to have some abstractions